File: 2699-count-the-number-of-fair-pairs/count-the-number-of-fair-pairs.py
class Solution:
    def countFairPairs(self, nums: List[int], lower: int, upper: int) -> int:


        count=0
        nums.sort()

        for i in range(len(nums)):
            y=lower-nums[i]
            x=upper-nums[i]
            count+= bisect_right(nums,x,i+1)-bisect_left(nums,y,i+1)
            print(count)


            
        return count

File: 2699-count-the-number-of-fair-pairs/count-the-number-of-fair-pairs.java
class Solution {
    private int BS(int low,int flag ,int target, int[] nums)
    {
        int high = nums.length;
        if(flag==0)
        {
            while(low < high)
            {
                int mid = low + (high - low) / 2;
                int val =nums[mid];
                if(val>=target)
                {
                   high=mid;
                }
                else{
                    low = mid +1;
                }
            }
        }
        else {
            while(low < high)
            {
                int mid = low + (high - low) / 2;
                int val =nums[mid];
                if(val<=target)
                {
                   low =mid+1;
                } else  {
                    high = mid;
                }

            }


        }
        return low;


    }
    public long countFairPairs(int[] nums, int lower, int upper) {
        Arrays.sort(nums);
        long res =0;
        for(int i=0;i<nums.length;i++)
        {
             int minReq = lower-nums[i];
             int maxReq = upper-nums[i];

             res+=BS(i+1,1,maxReq,nums) - BS(i+1,0,minReq,nums);
        }
        return res;
    }
}